File: backend/app/schemas/master_data.py
"""Shared schemas for master-data APIs and XLSX workflows."""

from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import Field

from app.schemas.base import ApiSchema


class SortOrder(str, Enum):
    ASC = "asc"
    DESC = "desc"


class ImportEntityType(str, Enum):
    DEPARTMENTS = "departments"
    SECTIONS = "sections"
    DOCUMENT_TYPES = "document-types"
    DOCUMENT_STATUSES = "document-statuses"
    VALIDATION_RULES = "validation-rules"


class ImportMode(str, Enum):
    CREATE_ONLY = "CREATE_ONLY"
    UPSERT = "UPSERT"


class ImportRowStatus(str, Enum):
    VALID = "VALID"
    INVALID = "INVALID"
    DUPLICATE = "DUPLICATE"


class MasterDataOption(ApiSchema):
    id: UUID
    code: str
    name: str
    is_active: bool


class MasterDataCount(ApiSchema):
    total: int = Field(ge=0)
    active: int = Field(ge=0)
    inactive: int = Field(ge=0)


class MasterDataOverview(ApiSchema):
    departments: MasterDataCount
    sections: MasterDataCount
    document_types: MasterDataCount
    document_statuses: MasterDataCount
    validation_rules: MasterDataCount


class ImportPreviewRow(ApiSchema):
    row_number: int = Field(ge=2)
    status: ImportRowStatus
    data: dict[str, Any]
    errors: list[str] = Field(default_factory=list)


class ImportPreviewResponse(ApiSchema):
    entity_type: ImportEntityType
    total_rows: int = Field(ge=0)
    valid_rows: int = Field(ge=0)
    invalid_rows: int = Field(ge=0)
    duplicate_rows: int = Field(ge=0)
    rows: list[ImportPreviewRow]
    warnings: list[str] = Field(default_factory=list)


class ImportConfirmResponse(ApiSchema):
    entity_type: ImportEntityType
    mode: ImportMode
    total_rows: int = Field(ge=0)
    created: int = Field(ge=0)
    updated: int = Field(ge=0)
    skipped: int = Field(ge=0)
    failed: int = Field(ge=0)

File: frontend/src/types/masterData.ts
export const masterDataEntityTypes = [
  'departments',
  'sections',
  'document-types',
  'document-statuses',
  'validation-rules',
] as const;

export type MasterDataEntityType = (typeof masterDataEntityTypes)[number];

export type SortOrder = 'asc' | 'desc';

export interface MasterDataAuditFields {
  createdAt: string;
  updatedAt: string;
  createdBy: string | null;
  updatedBy: string | null;
}

export interface MasterDataListParams {
  search?: string;
  isActive?: boolean;
  page: number;
  pageSize: number;
  sortBy?: string;
  sortOrder?: SortOrder;
}

export interface PaginatedData<TItem> {
  items: TItem[];
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
}

export interface MasterDataOption {
  id: string;
  code: string;
  name: string;
  isActive: boolean;
}

export interface OverviewCount {
  total: number;
  active: number;
  inactive: number;
}

export interface MasterDataOverview {
  departments: OverviewCount;
  sections: OverviewCount;
  documentTypes: OverviewCount;
  documentStatuses: OverviewCount;
  validationRules: OverviewCount;
}

export const importModes = ['CREATE_ONLY', 'UPSERT'] as const;
export type ImportMode = (typeof importModes)[number];

export const importRowStatuses = ['VALID', 'INVALID', 'DUPLICATE'] as const;
export type ImportRowStatus = (typeof importRowStatuses)[number];

export interface ImportPreviewRow {
  rowNumber: number;
  status: ImportRowStatus;
  data: Record<string, unknown>;
  errors: string[];
}

export interface ImportPreview {
  entityType: MasterDataEntityType;
  totalRows: number;
  validRows: number;
  invalidRows: number;
  duplicateRows: number;
  rows: ImportPreviewRow[];
  warnings: string[];
}

export interface ImportResult {
  entityType: MasterDataEntityType;
  mode: ImportMode;
  totalRows: number;
  created: number;
  updated: number;
  skipped: number;
  failed: number;
}

export interface BinaryDownload {
  blob: Blob;
  fileName: string | null;
}

export const masterDataPageSizes = [10, 20, 50, 100] as const;
